### PR TITLE
draft SSO docs

### LINKED ISSUE #7812: SSO in Admin UI
SSO will be implemented for the Admin UI.

- Document new auth flow
- Update any existing sign-on procedures, if necessary

Roadmap issue: https://cockroachlabs.atlassian.net/browse/CRDB-211
config and permissions docs: https://docs.google.com/document/d/1K7X1OBqlOAMsy12NlEOKu_wxXpu6TXa53L3bFJZYiIE/edit#
how-to steps: https://github.com/cockroachdb/cockroach/pull/52609

### LINKED ISSUE #7984: v19.2.11 Release Notes
Prep date: 10/5/20
Release date: 10/12/20
SHA: `e450c34dbf33745d889f41e1b1ea65197c1506fc`
Release Manager: @celiala 
Issue: https://github.com/cockroachdb/cockroach/issues/55256

[Docs To Do](https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/73105625/Release+process#Releaseprocess-16.Updatedocs):

- [ ] Run script to generate release notes
- [ ] Update docs site
- [ ] Edit release notes - Additions/changes/omissions to the release notes should be left in this issue's comments
- [ ] Once PR is merged / release notes are live, notify @meagan-g in the #releases Slack channel
- [ ] Email releases@ with link to published release notes
- [ ] Generate release notes issues (cc @lnhsingh)

### PR DESCRIPTION
**Pull request by taroface**
Fixes #7812.

Note: Admin UI screenshot is pending new login page & SSO button design.

### CODE REVIEW

**Comment by taroface** [comment]
@dhartunian Friendly ping for review!

**Comment by ericharmeling** [comment]
@taroface 

I don't think this closes #7984

**Review by dhartunian — COMMENTED** [review]
Apologies for the immense delay.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/8409#-:-MJ-gaRt7zZixtYV02Qk:b-xo7bz5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @taroface)

___
*[v20.2/sso.md, line 20 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-e08JBdk_JT9MGhHw:-MJ-e08JBdk_JT9MGhHx:b-hx2e96) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L20)):*
> ```Markdown
> 
> - An external OAuth 2.0 identity provider
> - A callback URL to redirect the user to the CockroachDB cluster
> ```

In theory the URL is "available" if the user can access the Admin UI since it's part of that server. It's a pre-requisite but wondering if it can be omitted.
___
*[v20.2/sso.md, line 30 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-eJjsClonEwvN4rVq:-MJ-eJjt4Gl18ihhfhFf:bqilv7h) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L30)):*
> ```Markdown
> 1. The user successfully authenticates with the provider, completing the OAuth flow.
> 1. The user is redirected to the CockroachDB cluster via a callback URL.
> 1. CockroachDB matches the information passed in the callback query to a registered SQL user.
> ```

Before this step, there's an exchange of a "code" for an "id_token". Not visible to user but wondering if it's worth pointing out. The callback query does not actually contain any user data, just the code.
___
*[v20.2/sso.md, line 43 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-enCTCsvgFiBW51-F:-MJ-enCTCsvgFiBW51-G:bs4gi2x) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L43)):*
> ```Markdown
> | `server.oidc_authentication.client_id` | '32789079457-g3hdfw8cbw85obi5cb525hsceaqf69unn.apps.googleusercontent.com'
> | `server.oidc_authentication.client_secret` | '6Q_jmRMgrPNOc_mN91boe-9EP'
> | `server.oidc_authentication.redirect_url` | 'https://localhost:8080/oidc/callback'
> ```

`/oidc/v1/callback`
___
*[v20.2/sso.md, line 51 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-eXOsCycfdCa8WznE:-MJ-eXOsCycfdCa8WznF:b-thd1yx) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L51)):*
> ```Markdown
> - `enabled` is a Boolean that enables or disables SSO.
> - `client_id` and `client_secret` are generated by the external identity provider.
> - `redirect_url` specifies the callback URL that redirects the user to CockroachDB after a successful authentication. This can be the address of a node in the cluster or the address of a load balancer that routes traffic to the nodes. The path **must** be appended with `/oidc/callback`.
> ```

Technically the path can be anything if it's on their proxy, the key is that it *routes* to the `/oidc/v1/callback` URL on the CRDB nodes. Probably not worth complicating this section though.

Either way, it's now version prefixed with `oidc/v1`.
___
*[v20.2/sso.md, line 53 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-esVhCYjY1SiHPd0l:-MJ-esVhCYjY1SiHPd0m:bytdmyq) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L53)):*
> ```Markdown
> - `redirect_url` specifies the callback URL that redirects the user to CockroachDB after a successful authentication. This can be the address of a node in the cluster or the address of a load balancer that routes traffic to the nodes. The path **must** be appended with `/oidc/callback`.
> - `provider_url` specifies the OAuth issuer identifier. Ensure that the URL does not have a terminating `/`. For more information, see the [OIDC specification](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig). Note that CockroachDB appends the required `/.well-known/openid-configuration` by default. You do not need to include it.
> - `scopes` is a space-delimited list of the [OAuth scopes](https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims) being requested for an Access Token. The `openid` scope is added by default and should not be included.
> ```

The `openid` scope is now required to be included, there's a validation that will check for its presence. Default value is just `"openid"` now
___
*[v20.2/sso.md, line 71 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-fztU5okd7_Um1Pwf:-MJ-fztU5okd7_Um1Pwg:bgxu833) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L71)):*
> ```Markdown
> 	Google OAuth 2.0 client details
> 
> 1. Add the URL for the Admin UI to the list of **Authorized Javascript origins**. On a local cluster, this will be `https://localhost:8080`. For details on how to determine this URL on a production cluster, see [Admin UI access](admin-ui-overview.html#admin-ui-access). 
> ```

I don't *think* this is necessary. Only the authorized redirect URIs are needed. I tested locally with removing the "localhost" from the JS origins and things still worked.
___
*[v20.2/sso.md, line 105 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-gSWQBGP9qERMXET2:-MJ-gSWQBGP9qERMXET3:b6dn9qa) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L105)):*
> ```Markdown
> 	{% include copy-clipboard.html %}
> 	~~~ sql
> 	> SET CLUSTER SETTING server.oidc_authentication.redirect_url = 'https://localhost:8080/oidc/callback';
> ```

`oidc/v1/callback`
___
*[v20.2/sso.md, line 114 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-gR6gCHfJPPgfR5i_:-MJ-gR6gCHfJPPgfR5ia:b-p0qr7g) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L114)):*
> ```Markdown
> 	{% include copy-clipboard.html %}
> 	~~~ sql
> 	> SET CLUSTER SETTING server.oidc_authentication.scopes = 'email';
> ```

`openid email`

**Review by taroface — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/8409)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[v20.2/sso.md, line 20 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-e08JBdk_JT9MGhHw:-MJXQkie4k_18njVa_cS:b-z0ttnj) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L20)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

In theory the URL is "available" if the user can access the Admin UI since it's part of that server. It's a pre-requisite but wondering if it can be omitted.
</blockquote></details>

OK - I think it makes sense to remove it.
___
*[v20.2/sso.md, line 30 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-eJjsClonEwvN4rVq:-MJYFou481iIMHYbUD3E:b-ubh2rx) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L30)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Before this step, there's an exchange of a "code" for an "id_token". Not visible to user but wondering if it's worth pointing out. The callback query does not actually contain any user data, just the code.
</blockquote></details>

I revised these steps a bit. Can you let me know if they're now correct?
___
*[v20.2/sso.md, line 43 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-enCTCsvgFiBW51-F:-MJXR-eZA-MyEEuRWpy0:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L43)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

`/oidc/v1/callback`
</blockquote></details>

Done.
___
*[v20.2/sso.md, line 51 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-eXOsCycfdCa8WznE:-MJXRJuL5Hqhall5TAdA:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L51)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Technically the path can be anything if it's on their proxy, the key is that it *routes* to the `/oidc/v1/callback` URL on the CRDB nodes. Probably not worth complicating this section though.

Either way, it's now version prefixed with `oidc/v1`.
</blockquote></details>

Done.
___
*[v20.2/sso.md, line 53 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-esVhCYjY1SiHPd0l:-MJXRl3c2BpaA-4rCnAL:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L53)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

The `openid` scope is now required to be included, there's a validation that will check for its presence. Default value is just `"openid"` now
</blockquote></details>

Done.
___
*[v20.2/sso.md, line 71 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-fztU5okd7_Um1Pwf:-MJXRq4REvzeEXdW3hlF:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L71)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I don't *think* this is necessary. Only the authorized redirect URIs are needed. I tested locally with removing the "localhost" from the JS origins and things still worked.
</blockquote></details>

Done.
___
*[v20.2/sso.md, line 105 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-gSWQBGP9qERMXET2:-MJXRsLd4562LwWr0ikV:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L105)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

`oidc/v1/callback`
</blockquote></details>

Done.
___
*[v20.2/sso.md, line 114 at r1](https://reviewable.io/reviews/cockroachdb/docs/8409#-MJ-gR6gCHfJPPgfR5i_:-MJXRwIz7JRp71FnQVvm:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cbab572a09a8d558af82078dbf6084859700505d/v20.2/sso.md#L114)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

`openid email`
</blockquote></details>

Done.

File: v20.2/sso.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+---
+title: Single Sign-on (Enterprise)
+summary: Implement single sign-on (SSO) for Admin UI access.
+toc: true
+---
+
+Single sign-on (SSO) allows a user to access the Admin UI in a secure cluster via an external identity provider. When SSO is configured and enabled, the [Admin UI login page](admin-ui-overview.html#admin-ui-access) will display an OAuth login button in addition to the password access option.
```
Comment (Amruta-Ranade) [inline v20.2/sso.md:7]: Can we specify which user we are referring to? CockroachDB user, SQL user, or Google user? And is the same user being referred to throughout the doc?

File: v20.2/sso.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+| `server.oidc_authentication.enabled` | true
+| `server.oidc_authentication.client_id` | '32789079457-g3hdfw8cbw85obi5cb525hsceaqf69unn.apps.googleusercontent.com'
+| `server.oidc_authentication.client_secret` | '6Q_jmRMgrPNOc_mN91boe-9EP'
+| `server.oidc_authentication.redirect_url` | 'https://localhost:8080/oidc/v1/callback'
+| `server.oidc_authentication.provider_url` | 'https://accounts.google.com'
+| `server.oidc_authentication.scopes` | 'openid profile email'
+| `server.oidc_authentication.claim_json_key` | 'email'
+| `server.oidc_authentication.principal_regex` | '^([^@]+)@[^@]+$'
+
+- `enabled` is a Boolean that enables or disables SSO.
```
Comment (Amruta-Ranade) [inline v20.2/sso.md:49]: We could add another column named "Description" and include these bullet points in the table.

**Review by taroface — COMMENTED** [review]
@dhartunian TFTR, and no worries! I had two follow-up questions for you.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/8409#-:-MJYksCV9gv3DczB2i9k:bx896ph)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Amruta-Ranade and @dhartunian)

___
*[v20.2/sso.md, line 7 at r3](https://reviewable.io/reviews/cockroachdb/docs/8409#-MHIaJ2CgyNOo0XbI12Y-r3-7:-MJYjX__7KCQ4d7G9sV-:bbofck8) ([raw file](https://github.com/cockroachdb/docs/blob/7f9e37a23f12e5dd6b1209c3e3979c958181da7f/v20.2/sso.md#L7)):*
<details><summary><i>Previously, Amruta-Ranade (Amruta Ranade) wrote…</i></summary><blockquote>

Can we specify which user we are referring to? CockroachDB user, SQL user, or Google user? And is the same user being referred to throughout the doc?
</blockquote></details>

I think this is just a CockroachDB user, since SSO can be used with any external identity provider. @dhartunian does that sound right?
___
*[v20.2/sso.md, line 49 at r3](https://reviewable.io/reviews/cockroachdb/docs/8409#-MHIaJ2CgyNOo0XbI12Y-r3-49:-MJYkkp63PZEMwIhg_tJ:bftcyz0) ([raw file](https://github.com/cockroachdb/docs/blob/7f9e37a23f12e5dd6b1209c3e3979c958181da7f/v20.2/sso.md#L49)):*
<details><summary><i>Previously, Amruta-Ranade (Amruta Ranade) wrote…</i></summary><blockquote>

We could add another column named "Description" and include these bullet points in the table.
</blockquote></details>

I tried this, but it made the table way too big and flowed off the page :(

**Review by dhartunian — APPROVED** [review]
> Reviewed 1 of 2 files at r2, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/8409)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Amruta-Ranade)

___
*[v20.2/sso.md, line 7 at r3](https://reviewable.io/reviews/cockroachdb/docs/8409#-MHIaJ2CgyNOo0XbI12Y-r3-7:-MLDJ3_v6E1jC6DWG6Ei:blfhk42) ([raw file](https://github.com/cockroachdb/docs/blob/7f9e37a23f12e5dd6b1209c3e3979c958181da7f/v20.2/sso.md#L7)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

I think this is just a CockroachDB user, since SSO can be used with any external identity provider. @dhartunian does that sound right?
</blockquote></details>

The SSO process logs you in as a specific SQL user but in this case they would not be using to actually execute any SQL so I think saying "CockroachDB User" makes sense.

**Pull request merged**